### DISCUSSION TITLE
LWG-3828 "Sync `intmax_t` and `uintmax_t` with C2x"

### DISCUSSION THREAD

**Discussion by CaseyCarter** [microsoft/STL#3457] [Ideas] [answered]
LWG-3828 notably makes it possible for implementations with 64-bit `intmax_t` to support 128-bit extended integral types without breaking ABI. This begs the question "Should the STL support Clang's 128-bit signed/unsigned integer types?"

Note that I believe this would also require a bit of work outside the STL. IIUC some operations result in calls to the compiler runtime, we'd have to arrange to link the proper clang runtimes or put the pertinent routines in, say, VCStartup.

**Comment by StephanTLavavej** [comment]
@strega-nil-ms would like to ask the MSVC compiler team ("nicely" :joy_cat: :zany_face:) to support 128-bit types.

@StephanTLavavej thinks that this is a non-goal of supporting non-Standard stuff and that we have enough work to do in the Standard itself.

**Reply by fsb4000** [reply]
> @strega-nil-ms would like to ask the MSVC compiler team ("nicely" 😹 🤪) to support 128-bit types.

+1

**Reply by CaseyCarter** [reply]
> @strega-nil-ms would like to ask the MSVC compiler team ("nicely" 😹 🤪) to support 128-bit types.

How is this related to the question? (I suspect the intent is "I/we think the STL should support 128-bit extended integer types only if both MSVC and Clang do" but it's nice to be explicit.)

> @StephanTLavavej thinks that this is a non-goal of supporting non-Standard stuff and that we have enough work to do in the Standard itself.

Extended integer types are not non-Standard; see the definition in [[basic.fundamental]/5](https://eel.is/c++draft/basic.fundamental#5.sentence-3). That said, they are certainly optional. We've ignored Clang's 128-bit types in the past arguing that there was no way to conformingly support them in the STL, LWG-3828 (together with WG21-P1467R9) correct that, making it possible for us to support extended integer types wider than `intmax_t` if we choose to. The point of this question is for us to reevaluate our position in light of this new information.

Personally, I don't see benefits commensurate with the work involved to justify adding this support, but I'm curious to see if anyone else does.

**Reply by strega-nil-ms** [reply]
I think that we should support them once MSVC supports them; the relation to the question is just that I am annoyed as a programmer than int128 isn't supported in MSVC.

**Reply by frederick-vs-ja** [reply]
FYI WG21-P3140 attempts to make some extended integer types mandatory. I don't think there's any existing platform on which `long long` is 128-bit or wider.

**Comment by StephanTLavavej** [comment, marked as answer]
There's no way this is happening in the foreseeable future.
